src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java

```diff
@@ -0,0 +1,282 @@
+      if (!Strings.isNullOrEmpty(kmsKeyName)) {
+        loadConfig.setDestinationEncryptionConfiguration(new EncryptionConfiguration().setKmsKeyName(kmsKeyName));
+      }
+
+      // Auto detect the schema if we're not given one, otherwise use the passed schema.
+      if (schema == null) {
+        LOG.info("No import schema provided, auto detecting schema.");
+        loadConfig.setAutodetect(true);
+      } else {
+        LOG.info("Using provided import schema '{}'.", schema.toString());
+      }
+
+      JobConfiguration config = new JobConfiguration();
+      config.setLoad(loadConfig);
+
+      // Get the dataset to determine the location
+      Dataset dataset =
+        bigQueryHelper.getRawBigquery().datasets().get(tableRef.getProjectId(), tableRef.getDatasetId()).execute();
+
+      JobReference jobReference =
+        bigQueryHelper.createJobReference(projectId, "direct-bigqueryhelper-import", dataset.getLocation());
+      Job job = new Job();
+      job.setConfiguration(config);
+      job.setJobReference(jobReference);
+
+      // Insert and run job.
+      bigQueryHelper.insertJobOrFetchDuplicate(projectId, job);
+
+      // Poll until job is complete.
+      waitForJobCompletion(bigQueryHelper.getRawBigquery(), projectId, jobReference);
+    }
+
+    /**
+     * This method is copied from BigQueryUtils#waitForJobCompletion for getting useful error message.
+     */
+    private static void waitForJobCompletion(Bigquery bigquery, String projectId,
+                                             JobReference jobReference) throws IOException, InterruptedException {
+
+      Sleeper sleeper = Sleeper.DEFAULT;
+      BackOff pollBackOff =
+        new ExponentialBackOff.Builder()
+          .setMaxIntervalMillis(BigQueryUtils.POLL_WAIT_INTERVAL_MAX_MILLIS)
+          .setInitialIntervalMillis(BigQueryUtils.POLL_WAIT_INITIAL_MILLIS)
+          .setMaxElapsedTimeMillis(BigQueryUtils.POLL_WAIT_MAX_ELAPSED_MILLIS)
+          .build();
+
+      // Get starting time.
+      long startTime = System.currentTimeMillis();
+      long elapsedTime;
+      boolean notDone = true;
+
+      // While job is incomplete continue to poll.
+      while (notDone) {
+        BackOff operationBackOff = new ExponentialBackOff.Builder().build();
+        Bigquery.Jobs.Get get = bigquery.jobs().get(projectId, jobReference.getJobId());
+
+        Job pollJob = ResilientOperation.retry(
+          ResilientOperation.getGoogleRequestCallable(get),
+          operationBackOff,
+          RetryDeterminer.RATE_LIMIT_ERRORS,
+          IOException.class,
+          sleeper);
+
+        elapsedTime = System.currentTimeMillis() - startTime;
+        LOG.debug("Job status ({} ms) {}: {}", elapsedTime, jobReference.getJobId(),
+                  pollJob.getStatus().getState());
+        if (pollJob.getStatus().getState().equals("DONE")) {
+          notDone = false;
+          if (pollJob.getStatus().getErrorResult() != null) {
```

Comment:
Is it possible for getErrorResult() to be non-null, but for errors to be empty or null? 

It's probably best to check for these scenarios to be safe. I think this check should be for the errors list, and we can have an 'else if' when getErrorResult() is not null

Reply:
As per javadoc`getErrorResult() != null` indicates an unsuccessful job. So this check should be here. However, the javadoc does not indicate any contract between these 2 methods. So we should do a null check for `getErrors` method before using it.